### PR TITLE
feat(fw): Changes required for Devnet-4 Tests

### PR DESCRIPTION
**Pull request by marioevz**
## 🗒️ Description

This PR can be best reviewed (and also merged via any method that is not Squash) by commit.

#### [Allow Bytes(), add sha256 method](https://github.com/ethereum/execution-spec-tests/pull/888/commits/a230a6c5726f5f1deb18849f522362284e05f7fe)

Allow instantiating an empty `Bytes` instance by calling it with no parameters, and also add a `sha256` method to all `Bytes` instances, which returns a `Hash` instance containing the sha256 digest.

#### [Allow instantiating Address with a label](https://github.com/ethereum/execution-spec-tests/pull/888/commits/31171347dca2111fe38e163655f338128e73d65c)

Adds keyword argument `label` to `Address` init method to allow it being instantiated with a label.

#### [Fix FixedSizeBytes comparison](https://github.com/ethereum/execution-spec-tests/pull/888/commits/20c04ef72307865d87221b947f06738e8e85badb)

Minor fix to `FixedSizeBytes` `__eq__` method to fix casting of the `other` element under comparison, and allow comparing against `None` without raising an exception.

#### [Add BlockchainTestEngine](https://github.com/ethereum/execution-spec-tests/pull/888/commits/69d447513e1af0d10ed081253b80c95e512e62a4)

Adds a `BlockchainTestEngine` spec which has all the same properties as `BlockchainTest` except that it only generates a blockchain test engine version fixture.

#### [Allow `nonce` in `fund_eoa`](https://github.com/ethereum/execution-spec-tests/pull/888/commits/a15156f338cb2b2572c59e2d3a469ae6b2341b28)

Allows specifying a starting nonce for the funded EOA, which could only be achieved during fill but not on execution.

#### [Allow Bytes in Initcode](https://github.com/ethereum/execution-spec-tests/pull/888/commits/6ef9589d2831dc9f4d0743bae5de5df81bf36360)

Minor fix to allow deploying arbitrary bytes in an Initcode.

## 🔗 Related Issues
None

## ✅ Checklist
- [x] All: Set appropriate labels for the changes.
- [x] All: Considered squashing commits to improve commit history.
- [x] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [x] Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been added to [converted-ethereum-tests.txt](/ethereum/execution-spec-tests/blob/main/converted-ethereum-tests.txt).
- [x] Tests: A PR with removal of converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been opened.
- [x] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

File: tests/prague/eip2935_historical_block_hashes_from_state/spec.py [hunk 1]
```diff
@@ -26,6 +28,6 @@ class Spec:
     """
 
     FORK_TIMESTAMP = 15_000
-    HISTORY_STORAGE_ADDRESS = 0x0AAE40965E6800CD9B1F4B05FF21581047E3F91E
+    HISTORY_STORAGE_ADDRESS = SystemContract.BLOCK_HISTORY
```
Comment (danceratopz) [inline tests/prague/eip2935_historical_block_hashes_from_state/spec.py:31]:
```suggestion
    HISTORY_STORAGE_ADDRESS = 0x0AAE40965E6800CD9B1F4B05FF21581047E3F91E
```

File: tests/prague/eip6110_deposits/spec.py [hunk 2]
```diff
@@ -24,4 +26,4 @@ class Spec:
     https://eips.ethereum.org/EIPS/eip-6110
     """
 
-    DEPOSIT_CONTRACT_ADDRESS = 0x00000000219AB540356CBB839CBE05303D7705FA
+    DEPOSIT_CONTRACT_ADDRESS = SystemContract.BEACON_DEPOSITS
```
Comment (danceratopz) [inline tests/prague/eip6110_deposits/spec.py:29]:
```suggestion
    DEPOSIT_CONTRACT_ADDRESS = 0x00000000219AB540356CBB839CBE05303D7705FA
```

File: tests/prague/eip7002_el_triggerable_withdrawals/spec.py [hunk 3]
```diff
@@ -30,8 +32,8 @@ class Spec:
     out.
     """
 
-    WITHDRAWAL_REQUEST_PREDEPLOY_ADDRESS = 0x00A3CA265EBCB825B45F985A16CEFB49958CE017
-    SYSTEM_ADDRESS = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFE
+    WITHDRAWAL_REQUEST_PREDEPLOY_ADDRESS = SystemContract.WITHDRAWAL_REQUESTS
```
Comment (danceratopz) [inline tests/prague/eip7002_el_triggerable_withdrawals/spec.py:35]:
```suggestion
    WITHDRAWAL_REQUEST_PREDEPLOY_ADDRESS = 0x00A3CA265EBCB825B45F985A16CEFB49958CE017
```

File: tests/prague/eip7251_consolidations/spec.py [hunk 4]
```diff
@@ -26,8 +28,8 @@ class Spec:
     https://eips.ethereum.org/EIPS/eip-7251#execution-layer
     """
 
-    CONSOLIDATION_REQUEST_PREDEPLOY_ADDRESS = 0x00B42DBF2194E931E80326D950320F7D9DBEAC02
-    SYSTEM_ADDRESS = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFE
+    CONSOLIDATION_REQUEST_PREDEPLOY_ADDRESS = SystemContract.CONSOLIDATION_REQUESTS
```
Comment (danceratopz) [inline tests/prague/eip7251_consolidations/spec.py:31]:
```suggestion
    CONSOLIDATION_REQUEST_PREDEPLOY_ADDRESS = 0x00B42DBF2194E931E80326D950320F7D9DBEAC02
```

**Review by danceratopz — APPROVED** [review]
Loving the precompile and system contract enums and the simplification this brings to forks.

I would suggest renaming this package from `ethereum_test_addresses` to `ethereum_addresses` as it's not really test specific.

Few comments below mainly regarding naming. At the cost of duplication, I would write the current version of system contract addresses in test `Spec` classes for ease of verification in the docs (see screenshot in comment below).

File: tests/cancun/eip4788_beacon_root/spec.py [hunk 5]
```diff
@@ -27,7 +28,7 @@ class Spec:
     https://eips.ethereum.org/EIPS/eip-4788#specification
     """
 
-    BEACON_ROOTS_ADDRESS = 0x000F3DF6D732807EF1319FB7B8BB8522D0BEAC02
+    BEACON_ROOTS_ADDRESS = SystemContract.BEACON_ROOT_HISTORY
```
[the remark this answers is not in the document]
Comment (marioevz) [reply tests/cancun/eip4788_beacon_root/spec.py:31]: I think this defeats the purpose of the library, which is to not have the same address in multiple places.

**Comment by marioevz** [comment]
Scrapped `ethereum_test_addresses` as it's unnecessary and contentious.

**Pull request merged**